core/modules/loaders.rs

```diff
@@ -82,6 +82,18 @@ pub trait ModuleLoader {
   ) -> Pin<Box<dyn Future<Output = Result<(), Error>>>> {
     async { Ok(()) }.boxed_local()
   }
+
+  /// Called when new v8 code cache is available for this module. Implementors
+  /// can store the provided code cache for future executions of the same module.
+  ///
+  /// It's not required to implement this method.
+  fn store_code_cache(
```

Comment:
Is there a reason this method is sync? Given that `prepare_module_load` and `load` are async maybe it would make sense to make it async as well? That way we wouldn't have to make blocking FS calls to store the code cache.

Reply:
Good point, will make it async.

How about the callback inside `op_eval_context`? `op_eval_context` is current sync; any issues with making it async?

Reply:
Actually, never mind. Given that `require` is sync I'd assume that we want to keep `op_eval_context` sync.

Reply:
Yeah, I think we should keep it sync - I think we can make API similar to Node's vm where you need to provide already loaded cache there.

Reply:
My current approach is to handle code cache outside of js, by using callbacks specified in `RuntimeOptions` by the embedder (see `core/ops_builtin_v8.rs`). Are you suggesting that we pass code cache objects through `evalContext` in javascript? That's going to make integration much more complex, and will also add more js/native transitions to read/write cache.

Reply:
Yes, that was my thinking - do you see immediate need to do it in `evalContext`? I think we should focus on validating the approach for proper ES modules and then we can expand handling of code cache to other places like `require()` and `Deno.core.evalContext`

Reply:
Yes, I'd like `require` to also be covered in this initial pass because it's making a big difference for cold starts for apps that use npm modules (which internally `require` many other npm modules).

Reply:
Hmmmm, then I guess there's no way around it. Then maybe we should have both sync and async version of providing code cache data? `require()` is already fully synchronous and block event loop, but it would be good if loading ES modules (either on startup or via dynamic import() wouldn't block it). WDYT?

Reply:
Yes, that's what the latest revision of this PR is doing. For ES modules, there is the async `ModuleLoader:: code_cache_ready` method.

For `require` there are 2 sync callbacks: to get and to save the code cache from/to the embedder.

Reply:
Sounds good, ship it!